File: contact-lookup.js
import fs from 'fs';

// Load admin contact mapping
let adminContactMapping = {};
try {
    const mappingData = fs.readFileSync('admin-contact-mapping.json', 'utf8');
    adminContactMapping = JSON.parse(mappingData);
} catch (error) {
    console.error('Không thể load admin contact mapping:', error.message);
}

/**
 * Tìm admin từ số điện thoại
 * @param {string} phone - Số điện thoại
 * @returns {object|null} - Admin info hoặc null nếu không tìm thấy
 */
export function findAdminByPhone(phone) {
    try {
        // Normalize phone number (remove spaces, dashes, etc.)
        const normalizedPhone = phone.replace(/[\s\-\(\)]/g, '');
        
        for (const [adminName, adminData] of Object.entries(adminContactMapping)) {
            if (adminData.phone && adminData.phone.replace(/[\s\-\(\)]/g, '') === normalizedPhone) {
                return {
                    name: adminData.name,
                    adminUrl: adminData.adminUrl,
                    facebookUrl: adminData.facebookUrl,
                    phone: adminData.phone,
                    matchType: 'phone'
                };
            }
        }
        
        return null;
    } catch (error) {
        console.error('Lỗi khi tìm admin bằng phone:', error.message);
        return null;
    }
}

/**
 * Tìm admin từ số tài khoản ngân hàng
 * @param {string} bankAccount - Số tài khoản
 * @returns {object|null} - Admin info hoặc null nếu không tìm thấy
 */
export function findAdminByBankAccount(bankAccount) {
    try {
        // Normalize bank account (remove spaces, dashes, etc.)
        const normalizedAccount = bankAccount.replace(/[\s\-]/g, '');
        
        for (const [adminName, adminData] of Object.entries(adminContactMapping)) {
            if (adminData.bankAccounts) {
                for (const [bankName, accountNumber] of Object.entries(adminData.bankAccounts)) {
                    if (accountNumber.replace(/[\s\-]/g, '') === normalizedAccount) {
                        return {
                            name: adminData.name,
                            adminUrl: adminData.adminUrl,
                            facebookUrl: adminData.facebookUrl,
                            phone: adminData.phone,
                            matchType: 'bank',
                            matchedBank: bankName,
                            matchedAccount: accountNumber
                        };
                    }
                }
            }
        }
        
        return null;
    } catch (error) {
        console.error('Lỗi khi tìm admin bằng bank account:', error.message);
        return null;
    }
}

/**
 * Detect và tìm admin từ text (phone hoặc bank account)
 * @param {string} text - Text chứa phone hoặc bank account
 * @returns {object|null} - Admin info hoặc null nếu không tìm thấy
 */
export function findAdminByContact(text) {
    // Regex patterns - cải thiện để lọc chính xác
    const phonePattern = /(?:^|\s)((?:0|\+84)[0-9]{8,10})(?=\s|$)/g;
    const bankAccountPattern = /(?:^|\s)(\d{10,20})(?=\s|$)/g;
    
    // Tìm phone numbers
    const phoneMatches = [...text.matchAll(phonePattern)];
    if (phoneMatches.length > 0) {
        for (const match of phoneMatches) {
            const phone = match[1]; // Lấy captured group
            const result = findAdminByPhone(phone);
            if (result) return result;
        }
    }
    
    // Tìm bank accounts
    const bankMatches = [...text.matchAll(bankAccountPattern)];
    if (bankMatches.length > 0) {
        for (const match of bankMatches) {
            const account = match[1]; // Lấy captured group
            const result = findAdminByBankAccount(account);
            if (result) return result;
        }
    }
    
    return null;
}

/**
 * Format thông tin admin contact để hiển thị
 * @param {object} admin - Admin data với contact info
 * @returns {string} - Formatted string
 */
export function formatAdminContactInfo(admin) {
    let info = `🕵️ **FB Real của: "${admin.name}"**\n`;
    info += `🎖 GDV này có bảo hiểm tại Checkscam.vn\n`;
    info += `🔗 ${admin.adminUrl}\n`;
    
    if (admin.facebookUrl) {
        info += `📘 Facebook: ${admin.facebookUrl}\n`;
    }
    
    if (admin.phone) {
        info += `📱 Phone: ${admin.phone}\n`;
    }
    
    if (admin.matchType === 'phone') {
        info += `✅ **Xác nhận qua SĐT**`;
    } else if (admin.matchType === 'bank') {
        info += `✅ **Xác nhận qua STK ${admin.matchedBank}**`;
    }
    
    return info;
}

// Test function
if (process.argv[1] && process.argv[1].endsWith('contact-lookup.js')) {
    console.log('🧪 Testing Contact Lookup...\n');
    
    // Test phone
    console.log('=== Test Phone ===');
    const phoneResult = findAdminByPhone('0763666222');
    if (phoneResult) {
        console.log('✅ Found by phone:');
        console.log(formatAdminContactInfo(phoneResult));
    } else {
        console.log('❌ Not found by phone');
    }
    
    // Test bank account
    console.log('\n=== Test Bank Account ===');
    const bankResult = findAdminByBankAccount('0491000133345');
    if (bankResult) {
        console.log('✅ Found by bank account:');
        console.log(formatAdminContactInfo(bankResult));
    } else {
        console.log('❌ Not found by bank account');
    }
    
    // Test mixed text
    console.log('\n=== Test Mixed Text ===');
    const mixedText = 'Anh em check giúp số này: 0763666222 và stk 0491000133345';
    const mixedResult = findAdminByContact(mixedText);
    if (mixedResult) {
        console.log('✅ Found in mixed text:');
        console.log(formatAdminContactInfo(mixedResult));
    } else {
        console.log('❌ Not found in mixed text');
    }
}

File: contact_lookup.py
#!/usr/bin/env python3
"""
Contact Lookup Module - Python Version
Tìm kiếm admin từ số điện thoại và số tài khoản ngân hàng
"""

import json
import re
import logging
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

# Load admin contact mapping
admin_contact_mapping = {}
try:
    with open('admin-contact-mapping.json', 'r', encoding='utf-8') as file:
        admin_contact_mapping = json.load(file)
except Exception as error:
    logger.error(f"Không thể load admin contact mapping: {error}")

def find_admin_by_phone(phone: str) -> Optional[Dict[str, Any]]:
    """
    Tìm admin từ số điện thoại
    """
    try:
        # Normalize phone number (remove spaces, dashes, etc.)
        normalized_phone = re.sub(r'[\s\-\(\)]', '', phone)
        
        for admin_name, admin_data in admin_contact_mapping.items():
            if admin_data.get('phone'):
                admin_phone = re.sub(r'[\s\-\(\)]', '', admin_data['phone'])
                if admin_phone == normalized_phone:
                    return {
                        'name': admin_data['name'],
                        'adminUrl': admin_data['adminUrl'],
                        'facebookUrl': admin_data.get('facebookUrl'),
                        'phone': admin_data['phone'],
                        'matchType': 'phone'
                    }
        
        return None
    except Exception as error:
        logger.error(f"Lỗi khi tìm admin bằng phone: {error}")
        return None

def find_admin_by_bank_account(bank_account: str) -> Optional[Dict[str, Any]]:
    """
    Tìm admin từ số tài khoản ngân hàng
    """
    try:
        # Normalize bank account (remove spaces, dashes, etc.)
        normalized_account = re.sub(r'[\s\-]', '', bank_account)
        
        for admin_name, admin_data in admin_contact_mapping.items():
            if admin_data.get('bankAccounts'):
                for bank_name, account_number in admin_data['bankAccounts'].items():
                    if re.sub(r'[\s\-]', '', account_number) == normalized_account:
                        return {
                            'name': admin_data['name'],
                            'adminUrl': admin_data['adminUrl'],
                            'facebookUrl': admin_data.get('facebookUrl'),
                            'phone': admin_data.get('phone'),
                            'matchType': 'bank',
                            'matchedBank': bank_name,
                            'matchedAccount': account_number
                        }
        
        return None
    except Exception as error:
        logger.error(f"Lỗi khi tìm admin bằng bank account: {error}")
        return None
def find_admin_by_contact(text: str) -> Optional[Dict[str, Any]]:
    """
    Detect và tìm admin từ text (phone hoặc bank account)
    """
    # Regex patterns - cải thiện để lọc chính xác
    phone_pattern = re.compile(r'(?:^|\s)((?:0|\+84)[0-9]{8,10})(?=\s|$)')
    bank_account_pattern = re.compile(r'(?:^|\s)(\d{10,20})(?=\s|$)')
    
    # Tìm phone numbers
    phone_matches = phone_pattern.findall(text)
    if phone_matches:
        for phone in phone_matches:
            result = find_admin_by_phone(phone)
            if result:
                return result
    
    # Tìm bank accounts
    bank_matches = bank_account_pattern.findall(text)
    if bank_matches:
        for account in bank_matches:
            result = find_admin_by_bank_account(account)
            if result:
                return result
    
    return None

def format_admin_contact_info(admin: Dict[str, Any]) -> str:
    """
    Format thông tin admin contact để hiển thị
    """
    info = f"🕵️ **FB Real của: \"{admin['name']}\"**\n"
    info += "🎖 GDV này có bảo hiểm tại Checkscam.vn\n"
    info += f"🔗 {admin['adminUrl']}\n"
    
    if admin.get('facebookUrl'):
        info += f"📘 Facebook: {admin['facebookUrl']}\n"
    
    if admin.get('phone'):
        info += f"📱 Phone: {admin['phone']}\n"
    
    if admin.get('matchType') == 'phone':
        info += "✅ **Xác nhận qua SĐT**"
    elif admin.get('matchType') == 'bank':
        info += f"✅ **Xác nhận qua STK {admin.get('matchedBank', '')}**"
    
    return info

# Test function
if __name__ == '__main__':
    print('🧪 Testing Contact Lookup...\n')
    
    # Test phone
    print('=== Test Phone ===')
    phone_result = find_admin_by_phone('0763666222')
    if phone_result:
        print('✅ Found by phone:')
        print(format_admin_contact_info(phone_result))
    else:
        print('❌ Not found by phone')
    
    # Test bank account
    print('\n=== Test Bank Account ===')
    bank_result = find_admin_by_bank_account('0491000133345')
    if bank_result:
        print('✅ Found by bank account:')
        print(format_admin_contact_info(bank_result))
    else:
        print('❌ Not found by bank account')
    
    # Test mixed text
    print('\n=== Test Mixed Text ===')
    mixed_text = 'Anh em check giúp số này: 0763666222 và stk 0491000133345'
    mixed_result = find_admin_by_contact(mixed_text)
    if mixed_result:
        print('✅ Found in mixed text:')
        print(format_admin_contact_info(mixed_result))
    else:
        print('❌ Not found in mixed text')